specs/SBOM_SPEC.md

```diff
@@ -0,0 +1,102 @@
+SBOMs attached with `cosign` should be retrievable in other tools, and vice-versa.
+
+This document focuses on the layout of an SBOM within an [OCI Image Manifest V1](https://github.com/opencontainers/image-spec/blob/master/manifest.md) object.
+
+This document does not prescribe any specific SBOM format.
+Multiple formats can be used, and the `mediaType` property should describe the format of a particular SBOM document.
+
+Multiple SBOMs may be "attached" to one image.
+Each SBOM may refer to the entire image, or to a specific part of that image.
+Exactly what the SBOM refers to is called the "scope" of that SBOM.
+
+SBOMs stored in an OCI registry are generally assumed to refer to other objects stored in that same registry.
+This document does not specify how these "links" are created.
+A naming convention or the [in-progress OCI `references` API](https://github.com/opencontainers/image-spec/issues/827) are viable options.
+
+This document does not specify how clients should behave When multiple SBOMs are present for an image.
+Clients may list all the SBOMs, or may provide tooling to filter based on SBOM type or scope.
+
+## Overall Layout
+
+An SBOM object is represented an [OCI Image Manifest V1](https://github.com/opencontainers/image-spec/blob/master/manifest.md).
+
+Each individual SBOM is represented as a `layer`, using a standard `descriptor`.
+The `layers` list is ordered, but no order is assumed or important for the SBOM documents.
+
+Here is an example manifest containing one SBOM, in the [SPDX](https://spdx.org) format:
+
+```json
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "application/vnd.oci.image.config.v1+json",
+    "size": 233,
+    "digest": "sha256:83bd5fb5b39f65f28e50a86d48fa79c07880befc292d92eebdc18531054b070c"
+  },
+  "layers": [
+    {
+      "mediaType": "text/spdx",
+      "size": 246,
+      "digest": "sha256:ed3ad03d3b87843b5419d7dce9d50a3e0f45554b2ba93bf378611cae6b450cff",
+    }
+  ]
+}
+```
+
+Multiple SBOMs may be attached, using multiple formats.
+This example shows two SBOMs, one in the SPDX format and one in the CycloneDX format:
+
+```json
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "application/vnd.oci.image.config.v1+json",
+    "size": 233,
+    "digest": "sha256:83bd5fb5b39f65f28e50a86d48fa79c07880befc292d92eebdc18531054b070c"
+  },
+  "layers": [
+    {
+      "mediaType": "text/spdx",
+      "size": 246,
+      "digest": "sha256:ed3ad03d3b87843b5419d7dce9d50a3e0f45554b2ba93bf378611cae6b450cff",
+    },
+    {
+      "mediaType": "application/vnd.cyclonedx",
+      "size": 462,
+      "digest": "sha256:e0851a4aa13657fc8dcd01e0e5e08cb817123ccb82e2c604b34f9ec9c1755e3f",
+    }
+  ]
+}
+```
+
+## MediaTypes
+
+The two main SBOM formats in use are [SPDX](https://spdx.org) and [CycloneDX](https://cyclonedx.org/).
+The `mediaTypes` for these should be indicated in the `descriptor` for each `layer`.
+
+The `mediaTypes` are:
```

Comment:
Both SPDX and CycloneDX can be representing in different formats, and different versions. Should the format be included as a subtype here?

eg.

* `application/vnd.cyclonedx+json`
* `application/vnd.cyclonedx+xml`

This would help clients who otherwise have to determine the format based on parsing along.

Reply:
Probably! I couldn't find anything other than text/spdx registered with IANA for that format. This makes sense for cyclonedx though.

Any ideas what would make sense for SPDX?

Reply:
These are in addition to the registered types, technically [Structured Syntax Suffixes](https://datatracker.ietf.org/doc/html/draft-ietf-appsawg-media-type-suffix-regs-02#section-3.1), so should be applicable to both. So I think:

* `text/spdx+json`
* `text/spdx+xml`

@stevespringett @coderpatros would be good to check for CycloneDX, and @kestewart for SPDX I think too.

Reply:
Confirming the CycloneDX media types that @garethr mentioned.

* [application/vnd.cyclonedx+xml](https://www.iana.org/assignments/media-types/application/vnd.cyclonedx+xml)
* [application/vnd.cyclonedx+json](https://www.iana.org/assignments/media-types/application/vnd.cyclonedx+json)

Both are officially registered with IANA.

Reply:
Thanks, added!

Reply:
@garethr Do you know if structured syntax suffixes are valid offical media types without being specified in the relevant RFC and media type registration? Text formats like SPDX tag/value have very different security considerations to something like XML.

Reply:
@jonjohnsonjr do you know?

Reply:
> Do you know if structured syntax suffixes are valid offical media types without being specified in the relevant RFC and media type registration?

I'm not sure I understand the question. Are you asking if you can make up suffixes that aren't registered? Or if you can use registered suffixes with arbitrary unregistered media types?

Reply:
@jonjohnsonjr whether or not you can use a registered suffix with a registered media type without registering the combination.

Reply:
https://datatracker.ietf.org/doc/html/rfc6839#section-2

>    Each of the structured syntax suffixes defined in this document is
>    appropriate for use when the media type identifies the semantics of
>    the protocol payload.  That is, knowing the semantics of the specific
>    media type provides for more specific processing of the content than
>    that afforded by generic processing of the underlying representation.
> 
>    At the same time, using the suffix allows receivers of the media
>    types to do generic processing of the underlying representation in
>    cases where
> 
>       they do not need to perform special handling of the particular
>       semantics of the exact media type, and
> 
>       there is no special knowledge needed by such a generic processor
>       in order to parse that underlying representation other than what
>       would be needed to parse any example of that underlying
>       representation.

Given that it's intended to be useful when received by a "generic processor", I would imagine it's fine to append a suffix to something without registering it, assuming that the content conforms to the definition of that suffix.

One example I have in mind is when displaying something -- it's useful to know if it's JSON or XML in order to correctly pretty-print some content, even if you don't know how to interpret the fields.

Also, with the vnd tree, you don't need to register a mediaType _anyway_, so that's at least one example of the suffix being valid even if the combination isn't registered.

https://datatracker.ietf.org/doc/html/rfc6838#section-3.2

>   While public exposure and review of media types to be registered in
>    the vendor tree are not required, using the media-types@iana.org
>    mailing list for review is encouraged, to improve the quality of
>    those specifications.  Registrations in the vendor tree may be
>    submitted directly to the IANA, where they will undergo Expert Review
>    [RFC5226] prior to approval.

Reply:
From what I've seen you have to register the subtype with the media type to make it official, for example, `svg+xml`:
https://www.iana.org/assignments/media-types/image/svg+xml

Reply:
Oh interesting @jonjohnsonjr . Three things come to mind:

1. I wonder if RDF would fall under `+xml` under this clause: "_in cases where they do not need to perform special handling of the particular semantics_"
2. `+yaml` is not defined in RFC 6839, would that require a specific registration?
3. I guess the real corner case is SPDX requiring an additional (new) domain-specific media type like, `+tagvalue`.

Reply:
YAML isn't registered as a media type or suffix, but probably should be. There's a thread here about doing so, but it looks like it never actually resolved: https://mailarchive.ietf.org/arch/msg/media-types/e9ZNC0hDXKXeFlAVRWxLCCaG9GI/

I would (boldly) suggest that it would be reasonable to use +yaml even if it's not registered, assuming the ecosystem understands what it means, but it would certainly be better if it were actually registered. I'm not sure who would be the best person to follow up with re: yaml registration.

The actual registries aren't the RFCs, but maintained here:
* https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml
* https://www.iana.org/assignments/media-types/media-types.xhtml

I'm not familiar enough to speak to the tagvalue thing -- is that specific to spdx?

Reply:
Exactly, while the doc can be expressed in json, yaml, rdf, etc tag value is like its own format: https://spdx.github.io/spdx-spec/appendix-IX-file-tags/#format

Reply:
Regarding using unregistered suffixes, like `+yaml`, this is from RFC6838 (noting a conflict is unlikely)

> "+suffix" constructs for as-yet unregistered structured syntaxes SHOULD NOT be used, given the possibility of conflicts with future suffix definitions.

And if it is not an official registered media type it should use the `x.` prefix to indicate it is in the unregistered tree as per RFC6838.

We did this for CycloneDX Protobuf format, for the above reasons, `application/x.vnd.cyclonedx+protobuf`.

Regardless, I don't think this is the place to be determining the SPDX media types. Just my 2c :)

Reply:
> And if it is not an official registered media type it should use the x. prefix to indicate it is in the unregistered tree as per RFC6838.

My reading is that this should basically never be used for public things:

https://datatracker.ietf.org/doc/html/rfc6838#section-3.4

>    Subtype names with "x." as the first facet may be used for types
>    intended exclusively for use in private, local environments.  Types
>    in this tree cannot be registered and are intended for use only with
>    the active agreement of the parties exchanging them.
> 
>    However, with the simplified registration procedures described above
>    for vendor and personal trees, it should rarely, if ever, be
>    necessary to use unregistered types.  Therefore, use of types in the
>    "x." tree is strongly discouraged.

Agreed that we SHOULD NOT use +yaml, but also we SHOULD register it, and that's not happening either :) 

> Regardless, I don't think this is the place to be determining the SPDX media types. Just my 2c :)

+1 -- I'm happy to spread my poor interpretation having skimmed a could RFCs, but this should be something the SPDX folks are responsible for.